drv: led: SCT2024 LED controller drivers

hubertmis opened the pull request:
Add a driver for SCT2024 LED controller and a sample demonstrating usage of this driver.

hubertmis committed:
drivers: led: sct2024 controller

Initial implementation of SCT2024 LED controller driver

Signed-off-by: Hubert Miś <hubert.mis@gmail.com>

hubertmis commented on samples/drivers/led/sct2024/src/main.c:
```diff
@@ -0,0 +1,57 @@
+		k_msleep(ON_OFF_DELAY_MS);
+	} while (true);
+
+	return 0;
```
This matches the signature of main() function and is common pattern in samples. Won't change

hubertmis commented on drivers/led/sct2024.c:
```diff
@@ -0,0 +1,215 @@
+	int ret = sct2024_spi_write(dev);
+
+	if (cfg->oe_pin.port) {
+		gpio_pin_set(cfg->oe_pin.port, cfg->oe_pin.pin, 1);
```
The purpose of using the OE pin is to avoid unnecessary SPI transfer when all LEDs are off. The state of the SCT2024 output pins is gated by OE so there previous LEDs state have no impact. The LEDs state is modified before OE pin is enabled again.

simonguinot commented on drivers/led/sct2024.c:
```diff
@@ -0,0 +1,219 @@
+		return -ENODEV;
+	}
+
+	if (!device_is_ready(cfg->la_pin.port)) {
```
gpio_is_ready_dt ^

simonguinot commented on drivers/led/sct2024.c:
```diff
@@ -0,0 +1,219 @@
+{
+	const struct sct2024_cfg *cfg = dev->config;
+
+	if (!device_is_ready(cfg->spi_dev)) {
```
spi_is_ready_dt ^

simonguinot commented on dts/bindings/led/sct,sct2024.yaml:
```diff
@@ -0,0 +1,19 @@
+    type: phandle-array
+  oe-gpios:
+    description: GPIO specifier for the OE (Output Enable) pin
+    type: phandle-array
```
Please, can you elaborate a bit more the description of this GPIOs ?

simonguinot commented on drivers/led/sct2024.c:
```diff
@@ -0,0 +1,219 @@
+	struct sct2024_data *data = dev->data;
+
+	if (cfg->oe_pin.port) {
+		bool all_leds_off = true;
```
Please move this definition at the top of the function.

simonguinot commented on drivers/led/sct2024.c:
```diff
@@ -0,0 +1,219 @@
+	struct sct2024_data *data = dev->data;
+	int led_index;
+
+	led_index = sct2024_get_led_index(dev, led);
```
There is a confusion here between LED node and controller indexes.

See https://github.com/zephyrproject-rtos/zephyr/pull/98635#pullrequestreview-3404759367 for a full explanation.

simonguinot commented on drivers/led/sct2024.c:
```diff
@@ -0,0 +1,219 @@
+		return -EIO;
+	}
+
+	if (cfg->oe_pin.port && !device_is_ready(cfg->oe_pin.port)) {
```
`gpio_is_ready_dt` ^

simonguinot reviewed (changes requested):
Hi @hubertmis,

Thanks for this driver. See my first review below.

In addition, I think that this driver may belong to the led-strip API.

simonguinot commented:
> Hi @simonguinot,
> 
> Thank you for the review.
> 
> Before I proceed with implementing changes I would like to confirm if this driver belongs to led or led_strip. To do so, let me provide a little context.
> 
> In the device I'm creating I have a LED matrix with current sinks controlled by SCT2024 and current sources controlled by transistors. For controlling the transistors I intended to use simple LED GPIO driver already available. For SCT2024 I created this driver. And then I was thinking to create a "led-matrix" driver implementing led.h API and using any LED driver to control matrix cols and another to control matrix rows. But maybe this matrix driver should be a led-strip or a display driver instead? Or maybe it should be a led driver using led-strip and led as its dependencies? What do you think?

Sorry @hubertmis but you lost me, especially on the part where you use both the SCT2024 and GPIO LED drivers to configure a display of LEDs :)

I am not sure which family the SCT2024 controller belongs to: led or led-strip. I'd say if the MOSI signal is consumed by the first controller of the chain and passed to the next, then led-strip is more suited. In addition, we already have a `led_strip_matrix` driver which can be used to build a display on the top of led-strip.

But I don't know the SCT2024 controller... Maybe it doesn't fit very well with the led-strip API and its RGB channels, with brightness control of each channel.  On the SCT2024 the ouput channels can be used to connect RGB LEDs or monochromatics LEDs. And there is no brightness granularity control via the MOSI signal. It is on or off. But it seems that the output enable pin allows to control the brightness via PWM but for all the channels of a controller.

So I don't know :)

hubertmis commented:
Let me attach schematics to clarify

<img width="1196" height="723" alt="image" src="https://github.com/user-attachments/assets/2829255d-0a96-4837-b9cc-452fe99b1e35" />

I believe the challenge here is that SCT2024 can be used as plain LED controller (like in my design) or, by chaining, as a LED strip controller. I don't think we can implement both APIs for a single device. Or can we?

fabiobaltieri commented on drivers/led/Kconfig.sct2024:
```diff
@@ -0,0 +1,15 @@
+	  communicates over SPI.
+
+	  If you say yes here, you will be able to control LEDs connected
+	  to the SCT2024 through this driver.
```
nit: feels like a very redundant comment :-) I would not bother

fabiobaltieri commented on drivers/led/sct2024.c:
```diff
@@ -0,0 +1,219 @@
+		return led_index;
+	}
+
+	data->led_bitmap[led_index / 16] &= ~(1 << (led_index % 16));
```
you can use `~BIT(led_index % 16)` here

simonguinot commented:
> SCT2024

Hi @hubertmis

I looked again at the datasheet of the SCT2024 controller and I think it is more compatible with the LED API than with led-strip (for the reasons I mentioned in a previous comment). Please, feel free to move on with this PR.

thedjnK commented:
> Let me attach schematics to clarify
> <img alt="image" width="1196" height="723" src="https://private-user-images.githubusercontent.com/21953319/511409279-2829255d-0a96-4837-b9cc-452fe99b1e35.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjQ0MDgzNTEsIm5iZiI6MTc2NDQwODA1MSwicGF0aCI6Ii8yMTk1MzMxOS81MTE0MDkyNzktMjgyOTI1NWQtMGE5Ni00ODM3LWI5Y2MtNDUyZmU5OWIxZTM1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTExMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMTI5VDA5MjA1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMxNGEyNzBhNzBhYWY3MGVmZTU0MGRmNzk4YzI4M2M1M2NlYTg3M2M1MmU0Njk0ZmE1ZmZhMTg0MjU5ZjExZjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.yEeycT8vNjSmCqrfTI3mOKzsg6r3P-BGgZsrENWMX90">
> 
> I believe the challenge here is that SCT2024 can be used as plain LED controller (like in my design) or, by chaining, as a LED strip controller. I don't think we can implement both APIs for a single device. Or can we?

 Yes, you can, you can have one device (MFD) which can be used as multiple devices and which driver is enabled is controlled by Kconfig - see (long) discussion in https://github.com/zephyrproject-rtos/zephyr/pull/96676#discussion_r2426742416

hubertmis commented:
> I think you mixed the changes between the two commits, looks like the updates for the first are in the second, can you take a look?

Right, a rebase mistake. Fixed

fabiobaltieri commented on drivers/led/sct2024.c:
```diff
@@ -0,0 +1,204 @@
+	 * iterations based on the greatest led index for the chain - detect the actual chain length
+	 */
+	BUILD_ASSERT(SCT2024_MAX_CHAIN_LENGTH == 1,
+		     "Driver currently supports only a single SCT2024 device in the chain");
```
these are evaluated at build time, bring it out of the function code, typically you'd put them either at the top or bottom of the file

fabiobaltieri commented on drivers/led/sct2024.c:
```diff
@@ -0,0 +1,204 @@
+{
+	const struct sct2024_cfg *cfg = dev->config;
+	struct sct2024_data *data = dev->data;
+	uint8_t buffer[SCT2024_MAX_CHAIN_LENGTH * 2] = {0};
```
does this need to be zero initialized? drop if not, does not look like it does to me

fabiobaltieri commented on drivers/led/sct2024.c:
```diff
@@ -0,0 +1,204 @@
+		return -EIO;
+	}
+
+	if (cfg->oe_pin.port && !gpio_is_ready_dt(&cfg->oe_pin)) {
```
drop `cfg->oe_pin.port &&`, `z_impl_device_is_ready` already has a null check

hubertmis committed:
sample: drivers: led: sct2024 controller

A sample demonstrating usage of a driver for SCT2024 LED controller.
This sample enables 16 leds one by one and then disables them one
by one in an inifite loop.

Signed-off-by: Hubert Miś <hubert.mis@gmail.com>

hubertmis commented:
@simonguinot , could you have another look?

JarmouniA commented on samples/drivers/led/sct2024/sample.yaml:
```diff
@@ -0,0 +1,8 @@
+  sample.drivers.led.sct2024:
+    filter: dt_compat_enabled("sct,sct2024")
+    tags: LED
+    depends_on: spi
```
```suggestion
```

JarmouniA commented on dts/bindings/led/sct,sct2024.yaml:
```diff
@@ -0,0 +1,23 @@
+properties:
+  la-gpios:
+    description:
+      GPIO specifier for the LA (Latch) pin.
```
please specify if it's expected to be active low or high.

JarmouniA commented on dts/bindings/led/sct,sct2024.yaml:
```diff
@@ -0,0 +1,23 @@
+    type: phandle-array
+  oe-gpios:
+    description:
+      GPIO specifier for the OE (Output Enable) pin.
```
ditto

simonguinot commented on drivers/led/sct2024.c:
```diff
@@ -0,0 +1,216 @@
+}
+
+static DEVICE_API(led, sct2024_led_api) = {
+	.set_brightness = sct2024_set_brightness,
```
Since you store the LED information, then you may want to implement the `->get_info` API as well. It  is almost free...

Note that an interesting optimization would be to also make the whole LED information thing optional, along with the `index` property. It costs some flash memory and it maybe useless for some applications.

simonguinot commented on drivers/led/sct2024.c:
```diff
@@ -0,0 +1,216 @@
+	const struct spi_buf_set tx = {
+		.buffers = &tx_buf,
+		.count = 1,
+	};
```
Please move the variables declarations at the beginning of the function.

simonguinot reviewed (commented):
Hi @hubertmis,

Please find my comments below.

hubertmis commented on drivers/led/sct2024.c:
```diff
@@ -0,0 +1,216 @@
+}
+
+static DEVICE_API(led, sct2024_led_api) = {
+	.set_brightness = sct2024_set_brightness,
```
I agree, but again, I would prefer to stick to basics now and introduce enhanced features later

thedjnK commented on drivers/led/CMakeLists.txt:
```diff
@@ -24,6 +24,7 @@ zephyr_library_sources_ifdef(CONFIG_MODULINO_BUTTONS_LEDS modulino_buttons_leds.
 zephyr_library_sources_ifdef(CONFIG_NCP5623 ncp5623.c)
 zephyr_library_sources_ifdef(CONFIG_PCA9533 pca9533.c)
```
fix commit title *`drivers:`

PerMac reviewed (changes requested):
It fails in twister when running with hw due to not being testable (and probably requiring extra hw like led strip).

PerMac commented on samples/drivers/led/sct2024/sample.yaml:
```diff
@@ -0,0 +1,9 @@
+tests:
+  sample.drivers.led.sct2024:
+    tags: led
+    platform_allow: nrf52840dk/nrf52840
```
or you can add some "fake" harness, e.g. "led" which will mean it is not testable in the current shape and will always be build-only

simonguinot reviewed (approved):
Thanks for this driver !
